### PR TITLE
k8s fixes

### PR DESCRIPTION
**Pull request by zeenix**
Same as https://github.com/kinvolk/habitat/pull/13 but against my unmerged parent branch for each review. :)

### CODE REVIEW

**Review by lilic — APPROVED** [review]
lgtm

**Comment by zeenix** [comment]
Thanks @LiliC . As @krnowak pointed out in #13, we are assuming that docker exporter will generate a specific tag and now that I checked the code, it will not produce the assumed tag if user passes `--no-tag-version` on the CLI. Moreover, even `:latest` tag can be disabled with a CLI arg. :( So I think it's best if k8s exporter ensure docker exporter used the same tags.

File: components/pkg-export-kubernetes/src/manifestjson.rs [hunk 1]
```diff
@@ -71,30 +68,29 @@ impl ManifestJson {
...
-    /// # Errors
-    ///
-    /// * Rendering from the template fails. This can only happen if the template has a syntax
-    /// error or the command-line arguments are incorrectly formatted. In both cases, it's a
-    /// most likely a programmer error if it happens.
-    // TODO: Implement TryInto trait instead when it's in stable std crate
-    pub fn into_string(&self) -> Result<String> {
+    fn into(self) -> String {
+        // The Result::expect() usage in this function is justied by the fact that errors can only
+        // come from crate programmer (e.g messed-up the manifest template or don't check the user
```
Comment (asymmetric) [inline components/pkg-export-kubernetes/src/manifestjson.rs]: Sorry to be a grammar freak again, but: I don't understand what the subject is in "e.g messed-up the manifest template or don't check the user input". So maybe reword it to something like "e.g. if the manifest template has syntax errors or user input is invalidat".

File: components/pkg-export-helm/src/main.rs [hunk 2]
```diff
@@ -54,7 +54,7 @@ fn main() {
 }
 
 fn export_for_cli_matches(ui: &mut UI, matches: &clap::ArgMatches) -> Result<()> {
-    let mut chart = Chart::new_for_cli_matches(ui, matches)?;
+    let chart = Chart::new_for_cli_matches(ui, matches)?;
```
Comment (asymmetric) [inline components/pkg-export-helm/src/main.rs:57]: It's good we can make something immutable, but one thing I don't understand: why was it necessary for it to be mutable before? And what changed in between?

File: components/pkg-export-kubernetes/src/manifest.rs [hunk 3]
```diff
@@ -129,7 +129,7 @@ impl Manifest {
 
     /// Generates the manifest as a string and writes it to `write`.
     pub fn generate(&mut self, write: &mut Write) -> Result<()> {
-        let out = ManifestJson::new(&self).into_string()?;
+        let out: String = ManifestJson::new(&self).into();
```
Comment (asymmetric) [inline components/pkg-export-kubernetes/src/manifest.rs:146]: A bit of a design smell with this IMO is that you end up having to use a type annotation anyway in the only place you use the `.into()`. I find `.into()` makes sense when you can avoid to do that cause the compiler manages to infer the type.

Another approach would be the [`ToString`](https://doc.rust-lang.org/alloc/string/trait.ToString.html#tymethod.to_string) trait, which you get by implementing the `Display` trait, which in turn could be handy as well.

My 2 cents.

File: components/pkg-export-helm/src/chart.rs [hunk 4]
```diff
@@ -125,7 +125,7 @@ impl<'a> Chart<'a> {
         }
     }
 
-    pub fn generate(&mut self) -> Result<()> {
+    pub fn generate(mut self) -> Result<()> {
```
Comment (asymmetric) [inline components/pkg-export-helm/src/chart.rs:126]: I wonder if this should be named differently, given how it consumes `self`. Maybe something like `convert`?

File: components/pkg-export-kubernetes/src/manifest.rs [hunk 5]
```diff
@@ -80,9 +80,20 @@ impl Manifest {
             PackageIdent::from_str(pkg_ident_str)?
         };
 
+        let tag = match pkg_ident.version {
+            Some(v) => {
+                match pkg_ident.release {
```
Comment (asymmetric) [inline components/pkg-export-kubernetes/src/manifest.rs]: I wonder if there's a way to do this with a chain of functions, something like: `pkg_ident.release.if_present(|r| format!(...))`.

Basically a way to inline the `match`.

**Review by asymmetric — CHANGES_REQUESTED** [review]
Mostly non-blocking comments.

File: components/pkg-export-helm/src/chart.rs (hunk 4 above)
Comment (zeenix) [reply components/pkg-export-helm/src/chart.rs:126]: I don't think so. Maybe when `TryInto` trait is in stable `std` crate, we should instead implement `TryInto<String>` here.

File: components/pkg-export-helm/src/main.rs (hunk 2 above)
Comment (zeenix) [reply components/pkg-export-helm/src/main.rs:57]: because of `chart.genarate()` not taking ownership and hence needing mutability IIRC. I only removed mut cause rustc informed me it's not needed anymore.

File: components/pkg-export-kubernetes/src/manifest.rs (hunk 3 above)
Comment (zeenix) [reply components/pkg-export-kubernetes/src/manifest.rs:146]: I don't think you decide to use Into trait or not base weather not a paricular usage of it will have to use type annotation. if the usage of 'out' afterwards would have made the type obvious to the compiler, we wouldn't need to annotate.

I did indeed go with implementing Display train actually but it felt wrong since that's mostly to create a user-visible string representation to show to user in messages and stuff. Although at that time, I hadn't decided that this method shouldn't return a Result. Now that I think of it, ToString() might be better indeed cause it's best to be explicit if possible but the docs discouraging implementations of it, makes me a bit vary of this trait.

File: components/pkg-export-kubernetes/src/manifestjson.rs (hunk 1 above)
Comment (zeenix) [reply components/pkg-export-kubernetes/src/manifestjson.rs]: sure.

File: components/pkg-export-kubernetes/src/manifest.rs (hunk 5 above)
Comment (zeenix) [reply components/pkg-export-kubernetes/src/manifest.rs]: for transforming the content of `Some`, there is Option::map() so i think it's doable. I'll check it out.

File: components/pkg-export-kubernetes/src/manifest.rs (hunk 3 above)
Comment (asymmetric) [reply components/pkg-export-kubernetes/src/manifest.rs:146]: Ask on IRC :)

File: components/pkg-export-kubernetes/src/manifest.rs (hunk 5 above)
Comment (zeenix) [reply components/pkg-export-kubernetes/src/manifest.rs]: i managed to get rid of the inner match (against release) but couldn't manage to do the same for the outer one w/o involving a string copy.

**Comment by zeenix** [comment]
Force push done!

File: components/pkg-export-docker/src/main.rs [hunk 6]
```diff
@@ -43,5 +43,5 @@ fn start(ui: &mut UI) -> Result<()> {
     let m = cli.get_matches();
     debug!("clap cli args: {:?}", m);
 
-    export_docker::export_for_cli_matches(ui, &m)
+    export_docker::export_for_cli_matches(ui, &m).map(|_| ())
```
Comment (asymmetric) [inline components/pkg-export-docker/src/main.rs:46]: It seems like nothing in the `start` function raises an error, so does its return type make sense? Maybe it should just have none, and then you woudn't have to use this `map` - which looks kinda weird, you're basically saying: iterate over every result, and completely ignore them.

File: components/pkg-export-kubernetes/src/manifest.rs [hunk 7]
```diff
@@ -91,9 +95,26 @@ impl Manifest {
...
-        let image = match matches.value_of("IMAGE_NAME") {
+        let image_name = match matches.value_of("IMAGE_NAME") {
             Some(i) => i.to_string(),
-            None => pkg_ident.origin + "/" + &pkg_ident.name + ":latest",
+            None => {
+                let (image_name, tag) = match image {
+                    Some(i) => {
+                        (
+                            i.name().to_owned(),
+                            i.tags().get(0).unwrap_or(&"latest".to_owned()).clone(),
```
Comment (asymmetric) [inline components/pkg-export-kubernetes/src/manifest.rs]: Out of curiosity, why do you need the `&`? Isn't a `&str` already a borrow?

Also, why do you need the `clone`? Rust is so hard sometimes 😭

File: components/pkg-export-kubernetes/src/manifest.rs [hunk 8]
```diff
@@ -91,9 +95,26 @@ impl Manifest {
         };
         let name = format!("{}-{}", pkg_ident.name, version_suffix);
 
-        let image = match matches.value_of("IMAGE_NAME") {
+        let image_name = match matches.value_of("IMAGE_NAME") {
```
Comment (asymmetric) [inline components/pkg-export-kubernetes/src/manifest.rs]: One way to kill another `match`, which I think would work:

```
value_of("IMAGE_NAME")
.map(|i| i.to_owned())
.or_else(<None block>)
```

Not sure if it's pushing it too far though 🤷‍♂️

File: components/pkg-export-kubernetes/src/manifest.rs (hunk 7 above)
Comment (krnowak) [reply components/pkg-export-kubernetes/src/manifest.rs]: `tags()` returns `&[String]` (reference to the slice of Strings). `get` on the slice returns something like `Option<&String>`.

`unwrap_or` for `Option<T>` looks like `fn unwrap_or(self, def: T) -> T`, so in our case where `T` is `&String`, `unwrap_or` is `fn unwrap_or(self, def: &String) -> &String`. We need to pass a reference to String to `unwrap_or`.

`&` in `&"latest".to_owned()` borrows a reference not to `"latest"` but to the result of `"latest".to_owned()`. `to_owned()` clones and returns `String` so that's why prepend `&` there, because we need `&String` here.

`unwrap_or` returns `&String`, so this stuff probably causes lifetime issues, so we need to clone it to get `String`. So if there was not tag at all we are cloning the cloned string. Whoop dee doo.

File: components/pkg-export-docker/src/main.rs (hunk 6 above)
Comment (krnowak) [reply components/pkg-export-docker/src/main.rs:46]: `export_to_cli_matches` used to return `Result<()>`, now it returns `Result<DockerImage>`, so we need to turn it back to `Result<()>`. This can be done that way or with `.and(Ok(()))`, or with:
```
export_docker::export_for_cli_matches(ui, &m)?;
Ok(())
```

File: components/pkg-export-kubernetes/src/manifest.rs (hunk 8 above)
Comment (krnowak) [reply components/pkg-export-kubernetes/src/manifest.rs]: I wonder what's wrong with `match`…

Or are we playing perlgolf? Err, I mean - rustgolf?

File: components/pkg-export-kubernetes/src/manifest.rs (hunk 7 above)
Comment (krnowak) [reply components/pkg-export-kubernetes/src/manifest.rs]: The double clone probably could be avoided with:
```
match i.tags().get(0) {
    Some(ref s) => s.clone(),
    None => "latest".to_owned(),
}
```

Comment (zeenix) [reply components/pkg-export-kubernetes/src/manifest.rs]: Thanks for explaining @krnowak . But there is no actual double clone here. `&"latest.to_owned()` is only happening if value is None, so your match statement actually does exactly the same.

File: components/pkg-export-kubernetes/src/manifest.rs (hunk 8 above)
Comment (zeenix) [reply components/pkg-export-kubernetes/src/manifest.rs]: nothing wrong with `match` really and it's one of the coolest thing after sliced bread but I've seen rust devs prefering `Option` & `Result` method chaining over match and usually it makes things very compact so why not. However, there is no reason to be go on a "must kill all match statements" spree. :)

File: components/pkg-export-docker/src/main.rs (hunk 6 above)
Comment (zeenix) [reply components/pkg-export-docker/src/main.rs:46]: @asymmetric what @krnowak said and 'map' on Result and Option doesn't iterate, just maps one content to another.

I like the `and(Ok())` better than the explicit separate return.

File: components/pkg-export-kubernetes/src/manifest.rs (hunk 7 above)
Comment (asymmetric) [reply components/pkg-export-kubernetes/src/manifest.rs]: Asked on IRC, this was presented as a better solution:

`i.tags().get(0).cloned().unwrap_or_else(|| "latest".to_owned())`.

2 changes:

* [`cloned()`](https://doc.rust-lang.org/std/option/enum.Option.html#method.cloned) - From `Option<&T>` to `Option<T>`. Kind of a syntactic sugar to remove the last `clone()`. 
* `unwrap_or_else` instead of `unwrap_or` - since it's the body of a function, it's only executed if needed. Previously, the `&str` was being cloned every time.

File: components/pkg-export-kubernetes/src/manifest.rs (hunk 8 above)
Comment (asymmetric) [reply components/pkg-export-kubernetes/src/manifest.rs]: Agreed, I've also had the impression that chaining is more idiomatic - part of the charm of rust are these functional programming inspired bits 👅

File: components/pkg-export-kubernetes/src/manifest.rs (hunk 7 above)
Comment (krnowak) [reply components/pkg-export-kubernetes/src/manifest.rs]: This solution sounds nice. Better than current one, because `"latest".to_owned()` is executed every time now - it is just a parameter to a function call.

Comment (asymmetric) [reply components/pkg-export-kubernetes/src/manifest.rs]: Forgot to mention that `cloned()` also allows us to remove the ugly `&` in `&"latest"` 👍

File: components/pkg-export-docker/src/main.rs (hunk 6 above)
Comment (zeenix) [reply components/pkg-export-docker/src/main.rs:46]: After changing, I'm no longer sure it looks better:

-    export_docker::export_for_cli_matches(ui, &m).map(|_| ())
+    export_docker::export_for_cli_matches(ui, &m).and(Ok(()))

meh. :)

Comment (asymmetric) [reply components/pkg-export-docker/src/main.rs:46]: I think the second one conveys the meaning better.

**Review by asymmetric — APPROVED** [review]

File: components/pkg-export-docker/src/main.rs (hunk 6 above)
Comment (zeenix) [reply components/pkg-export-docker/src/main.rs:46]: hmm... maybe but the ugliness of `Ok(())` annoys me more. :)

Comment (asymmetric) [reply components/pkg-export-docker/src/main.rs:46]: It's beautiful.
